zcncore/transaction_query_base.go

```diff
@@ -140,57 +140,81 @@ func (tq *TransactionQuery) checkHealth(ctx context.Context, host string) error
 	errs := r.Wait()
 
 	if len(errs) > 0 {
+		if errors.Is(errs[0], context.DeadlineExceeded) {
+			return context.DeadlineExceeded
+		}
 		tq.Lock()
 		tq.offline[host] = true
 		tq.Unlock()
-
-		if len(tq.offline) >= tq.max {
-			return ErrNoOnlineSharders
-		}
+		return ErrSharderOffline
 	}
-
 	return nil
 }
 
-// randOne random one health sharder
-func (tq *TransactionQuery) randOne(ctx context.Context) (string, error) {
-
-	randGen := rand.New(rand.NewSource(time.Now().UnixNano()))
-	for {
-
-		// reset selected if all sharders were selected
-		if len(tq.selected) >= tq.max {
-			tq.selected = make(map[string]bool)
+// getRandomSharder returns a random healthy sharder
+func (tq *TransactionQuery) getRandomSharder(ctx context.Context) (string, error) {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	for i := 0; i < len(tq.sharders); i += tq.numShardersToBatch {
```

Comment:
We will lost the randomness as we always start from the same 3 sharders. We can copy the sharders, shuffle them, and keep all the rest the same but use the shuffled sharders.

Reply:
Thanks for pointing this out. Fixed this in next commit.